pkg/operator/storageclasscontroller/vmware.go

```diff
@@ -89,33 +104,135 @@ func (v *storagePolicyAPI) getDefaultDatastore(ctx context.Context) (*mo.Datasto
 	return &dsMo, nil
 }
 
+func (v *storagePolicyAPI) getDatastore(ctx context.Context, dcName string, dsName string) (*mo.Datastore, error) {
+	vmClient := v.vcenterApiConnection.Client
+	finder := find.NewFinder(vmClient.Client, false)
+
+	dc, err := finder.Datacenter(ctx, dcName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access datacenter %s: %s", dcName, err)
+	}
+
+	finder.SetDatacenter(dc)
+	ds, err := finder.Datastore(ctx, dsName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to access datastore %s: %s", dsName, err)
+	}
+
+	var dsMo mo.Datastore
+	pc := property.DefaultCollector(dc.Client())
+	properties := []string{DatastoreInfoProperty, SummaryProperty}
+	err = pc.RetrieveOne(ctx, ds.Reference(), properties, &dsMo)
+	if err != nil {
+		return nil, fmt.Errorf("error getting properties of datastore %s: %v", dsName, err)
+	}
+	return &dsMo, nil
+}
+
+func (v *storagePolicyAPI) createZonalStoragePolicy(ctx context.Context) (string, error) {
+	var err error
+	failureDomains := v.infra.Spec.PlatformSpec.VSphere.FailureDomains
+	if len(failureDomains) <= 1 {
+		return "", fmt.Errorf("cluster has only one failure domain defined")
+	}
+
+	var aggregatedErrors []error
+
+	for _, failureDomain := range failureDomains {
+		dataCenter := failureDomain.Topology.Datacenter
+		dataStore := failureDomain.Topology.Datastore
+		err := v.attachTags(ctx, dataCenter, dataStore)
+		if err != nil {
+			aggregatedErrors = append(aggregatedErrors, err)
+		}
+	}
+
+	if len(aggregatedErrors) > 0 {
+		return v.policyName, errors.NewAggregate(aggregatedErrors)
+	}
+
+	// If we already created storage policy, we should return it
+	if v.policyCreated {
+		return v.policyName, nil
+	}
+
+	err = v.createStorageProfile(ctx)
+	if err != nil {
+		return v.policyName, fmt.Errorf("error create storage policy profile %s: %v", v.policyName, err)
+	}
+
+	return v.policyName, nil
+}
+
+func (v *storagePolicyAPI) attachTags(ctx context.Context, dcName, dsName string) error {
+	ds, err := v.getDatastore(ctx, dcName, dsName)
+	if err != nil {
+		return fmt.Errorf("unable to fetch datastore %s: %v", dsName, err)
+	}
+
+	// skip creation of tags on a datastore if it already has been tagged.
+	if v.policyCreated && v.checkForTagOnDatastore(ctx, ds) {
+		return nil
+	}
+
+	err = v.createOrUpdateTag(ctx, ds)
+	if err != nil {
+		return fmt.Errorf("error creating or updating tag %s: %v", v.tagName, err)
```

Comment:
is the datastore name already part of `err` from `createOrUpdateTag` / `tagManager.AttachTag` inside? If not, it could be useful for a cluster admin in a multi-datastore cluster.

Reply:
fixed